public/app/plugins/panel/heatmap-new/utils.ts

```diff
@@ -460,3 +468,64 @@ export const countsToFills = (u: uPlot, seriesIdx: number, palette: string[]) =>
 
   return indexedFills;
 };
+
+export const getHeatmapFields = (dataFrame: DataFrame): Array<Field | undefined> => {
+  const xField: Field | undefined = dataFrame.fields.find((f) => f.name === 'xMin');
+  const yField: Field | undefined = dataFrame.fields.find((f) => f.name === 'yMin');
+  const countField: Field | undefined = dataFrame.fields.find((f) => f.name === 'count');
+
+  return [xField, yField, countField];
+};
+
+export const getHeatmapArrays = (dataFrame: DataFrame): Array<number[] | undefined> => {
+  const [xField, yField, countField] = getHeatmapFields(dataFrame);
+  return [xField?.values.toArray(), yField?.values.toArray(), countField?.values.toArray()];
+};
+
+export const getDataMapping = (
+  heatmapData: HeatmapData,
+  origData: DataFrame,
+  options?: DataMapOptions
+): Array<number[] | null> => {
+  const [fxs, fys, fcounts] = getHeatmapFields(heatmapData.heatmap!);
+  const xos: number[] | undefined = origData.fields.find((f: Field) => f.type === 'time')?.values.toArray();
+  const yos: number[] | undefined = origData.fields.find((f: Field) => f.type === 'number')?.values.toArray();
+
+  if (heatmapData.heatmap?.meta?.type !== DataFrameType.HeatmapScanlines) {
+    return [null];
+  }
+
+  if (fxs && fys && fcounts && xos && yos) {
+    const mapping: Array<number[] | null> = new Array(heatmapData.xBucketCount! * heatmapData.yBucketCount!).fill(null);
+    const xsmin = options?.xMin ?? fxs.state?.calcs?.min ?? fxs.state?.range?.min ?? fxs.values.get(0);
+    const ysmin = options?.yMin ?? fys.state?.calcs?.min ?? fys.state?.range?.min ?? fys.values.get(0);
+    const xsmax =
+      options?.xMax ?? fxs.values.get(fxs.values.length - 1) + heatmapData.xBucketSize! * heatmapData.xBucketCount!;
+    const ysmax =
+      options?.yMax ?? fys.values.get(fys.values.length - 1) + heatmapData.yBucketSize! * heatmapData.yBucketCount!;
+    xos.forEach((xo: number, i: number) => {
+      const yo = yos[i];
+      const xBucketIdx = Math.floor(incrRoundDn(incrRoundUp((xo - xsmin) / heatmapData.xBucketSize!, 1e-7), 1e-7));
+      const yBucketIdx = Math.floor(incrRoundDn(incrRoundUp((yo - ysmin) / heatmapData.yBucketSize!, 1e-7), 1e-7));
+
+      if (xo < xsmin || xo > xsmax) {
+        return;
+      }
+
+      if (yo < ysmin || yo > ysmax) {
+        return;
+      }
+
+      const index = xBucketIdx * heatmapData.yBucketCount! + yBucketIdx;
+      const count = fcounts.values.get(index);
+      if (count > 0 || !options?.requireCount) {
```

Comment:
what is requireCount?  do we really need it?  

when checking the final results, we can easily do:
```
if (mapping[index]?.length) {
 ...
}
```

Reply:
`requireCount` essentially says "if there are no values in this bucket, don't try to process any values into this bucket. It ensures that (for example) exemplars only fall into buckets where there was a rendered cell on the heatmap.